File: elixir/fileio.ex
defmodule Main do
  @read_chunk_size 65536  # 64 KiB read buffer
  @chunk_size 65536       # chunk size for counting

  def main() do
    [file_count_str, n_str] = System.argv()
    file_count = String.to_integer(file_count_str)
    n = String.to_integer(n_str)

    # Pre-create content string with n '1' characters
    content = :binary.copy("1", n)

    # Write phase - write entire content at once
    write_files(file_count, content)

    # Read phase - read in chunks and count total bytes
    total_bytes = read_files(file_count)

    # Count chunks of 65536 bytes
    total_chunks = div(total_bytes, @chunk_size)
    IO.puts(total_chunks)

    # Cleanup
    cleanup_files(file_count)
  end

  defp write_files(file_count, content) do
    for i <- 0..(file_count - 1) do
      filename = "file#{i}"
      File.write!(filename, content)
    end
  end

  defp read_files(file_count) do
    Enum.reduce(0..(file_count - 1), 0, fn i, acc ->
      filename = "file#{i}"
      total = read_file_chunks(filename, 0)
      acc + total
    end)
  end

  defp read_file_chunks(filename, acc) do
    {:ok, device} = File.open(filename, [:read, :binary])
    total = read_chunks_loop(device, acc)
    File.close(device)
    total
  end

  defp read_chunks_loop(device, acc) do
    case IO.read(device, @read_chunk_size) do
      {:error, _} -> acc
      :eof -> acc
      data when is_binary(data) ->
        read_chunks_loop(device, acc + byte_size(data))
    end
  end

  defp cleanup_files(file_count) do
    for i <- 0..(file_count - 1) do
      filename = "file#{i}"
      File.rm!(filename)
    end
  end
end

File: hs/fileio.hs
{-# LANGUAGE BangPatterns #-}

import System.Environment (getArgs)
import System.IO
import qualified Data.ByteString as BS
import Data.ByteString.Builder (byteString, toLazyByteString)
import qualified Data.ByteString.Lazy as LBS
import System.Directory (removeFile)

main :: IO ()
main = do
  [fileCountStr, nStr] <- getArgs
  let !fileCount = read fileCountStr :: Int
      !n = read nStr :: Int

  -- Pre-create content with n '1' characters
  let !content = BS.replicate n 49  -- 49 is ASCII for '1'

  -- Write phase
  writeFiles fileCount content

  -- Read phase - read in chunks and count total bytes
  !totalBytes <- readFiles fileCount

  -- Count chunks of 65536 bytes
  let !totalChunks = totalBytes `div` 65536
  print totalChunks

  -- Cleanup
  cleanupFiles fileCount

writeFiles :: Int -> BS.ByteString -> IO ()
writeFiles fileCount content = do
  let filenames = ["file" ++ show i | i <- [0..fileCount-1]]
  mapM_ (\fname -> BS.writeFile fname content) filenames

readFiles :: Int -> IO Integer
readFiles fileCount = do
  let filenames = ["file" ++ show i | i <- [0..fileCount-1]]
  total <- sum <$> mapM readFileSize filenames
  return $! fromIntegral total

readFileSize :: FilePath -> IO Int
readFileSize filename = do
  handle <- openBinaryFile filename ReadMode
  total <- readChunks handle 0
  hClose handle
  return total
  where
    readChunks :: Handle -> Int -> IO Int
    readChunks h !acc = do
      chunk <- BS.hGet h 65536
      if BS.null chunk
        then return acc
        else readChunks h (acc + BS.length chunk)

cleanupFiles :: Int -> IO ()
cleanupFiles fileCount = do
  let filenames = ["file" ++ show i | i <- [0..fileCount-1]]
  mapM_ removeFile filenames